File: hooks.py
from __future__ import annotations

from typing import Any, Awaitable, Callable

PreHook = Callable[[str, dict], Awaitable[None]]
PostHook = Callable[[str, dict, str], Awaitable[None]]

_pre_hooks: list[PreHook] = []
_post_hooks: list[PostHook] = []


def register_pre(fn: PreHook) -> None:
    _pre_hooks.append(fn)


def register_post(fn: PostHook) -> None:
    _post_hooks.append(fn)


async def pre(tool_name: str, args: dict) -> None:
    for fn in _pre_hooks:
        await fn(tool_name, args)


async def post(tool_name: str, args: dict, result: str) -> None:
    for fn in _post_hooks:
        await fn(tool_name, args, result)

File: src/hooks.ts
export type PreHook = (toolName: string, args: Record<string, unknown>) => Promise<void>;
export type PostHook = (toolName: string, args: Record<string, unknown>, result: string) => Promise<void>;

const _preHooks: PreHook[] = [];
const _postHooks: PostHook[] = [];

export function registerPre(fn: PreHook): void {
  _preHooks.push(fn);
}

export function registerPost(fn: PostHook): void {
  _postHooks.push(fn);
}

export async function pre(toolName: string, args: Record<string, unknown>): Promise<void> {
  for (const fn of _preHooks) {
    await fn(toolName, args);
  }
}

export async function post(toolName: string, args: Record<string, unknown>, result: string): Promise<void> {
  for (const fn of _postHooks) {
    await fn(toolName, args, result);
  }
}
